<h1>UK ROAD ACCIDENT DATA ANALYTICS PROJECT</h1>
<hr>
<h3>Analyst : Jeremy Bryan T. Villanueva</h3>

<h2>Libraries</h2>

In [1]:
#import libraries
import numpy as np
import pandas as pd
import warnings
warnings.filterwarnings('ignore')

<h2>Converting Datasets into DataFrame</h2>

In [2]:
accident = pd.read_csv('dataset\\uk_road_accident.csv')
accident

Index Accident_Severity Accident Date   Latitude  \
0       200701BS64157           Serious      5/6/2019  51.506187   
1       200701BS65737           Serious      2/7/2019  51.495029   
2       200701BS66127           Serious    26-08-2019  51.517715   
3       200701BS66128           Serious    16-08-2019  51.495478   
4       200701BS66837            Slight      3/9/2019  51.488576   
...               ...               ...           ...        ...   
660674  201091NM01760            Slight    18-02-2022  57.374005   
660675  201091NM01881            Slight    21-02-2022  57.232273   
660676  201091NM01935            Slight    23-02-2022  57.585044   
660677  201091NM01964           Serious    23-02-2022  57.214898   
660678  201091NM02142           Serious    28-02-2022  57.575210   

                   Light_Conditions           District Area  Longitude  \
0             Darkness - lights lit  Kensington and Chelsea  -0.209082   
1                          Daylight  Kensington and Chelsea  -0.173647   
2       Darkness - lighting unknown  Kensington and Chelsea  -0.210215   
3                          Daylight  Kensington and Chelsea  -0.202731   
4             Darkness - lights lit  Kensington and Chelsea  -0.192487   
...                             ...                     ...        ...   
660674                     Daylight                Highland  -3.467828   
660675       Darkness - no lighting                Highland  -3.809281   
660676                     Daylight                Highland  -3.862727   
660677       Darkness - no lighting                Highland  -3.823997   
660678                     Daylight                Highland  -3.895673   

        Number_of_Casualties  Number_of_Vehicles Road_Surface_Conditions  \
0                          1                   2                     Dry   
1                          1                   2             Wet or damp   
2                          1                   3                     Dry   
3                          1                   4                     Dry   
4                          1                   2                     Dry   
...                      ...                 ...                     ...   
660674                     2                   1                     Dry   
660675                     1                   1            Frost or ice   
660676                     1                   3            Frost or ice   
660677                     1                   2             Wet or damp   
660678                     1                   1             Wet or damp   

                 Road_Type Urban_or_Rural_Area     Weather_Conditions  \
0       Single carriageway               Urban     Fine no high winds   
1       Single carriageway               Urban  Raining no high winds   
2                      NaN               Urban                    NaN   
3       Single carriageway               Urban     Fine no high winds   
4                      NaN               Urban                    NaN   
...                    ...                 ...                    ...   
660674  Single carriageway               Rural     Fine no high winds   
660675  Single carriageway               Rural     Fine no high winds   
660676  Single carriageway               Rural     Fine no high winds   
660677  Single carriageway               Rural     Fine no high winds   
660678    Dual carriageway               Rural  Snowing no high winds   

                                Vehicle_Type  
0                                        Car  
1                                        Car  
2                      Taxi/Private hire car  
3       Bus or coach (17 or more pass seats)  
4                              Other vehicle  
...                                      ...  
660674                                   Car  
660675                                   Car  
660676                                   Car  
660677                 Motorcycle over 500cc  
660

<h2>DataFrame Info</h2>

In [3]:
accident.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 660679 entries, 0 to 660678
Data columns (total 14 columns):
 #   Column                   Non-Null Count   Dtype  
---  ------                   --------------   -----  
 0   Index                    660679 non-null  object 
 1   Accident_Severity        660679 non-null  object 
 2   Accident Date            660679 non-null  object 
 3   Latitude                 660654 non-null  float64
 4   Light_Conditions         660679 non-null  object 
 5   District Area            660679 non-null  object 
 6   Longitude                660653 non-null  float64
 7   Number_of_Casualties     660679 non-null  int64  
 8   Number_of_Vehicles       660679 non-null  int64  
 9   Road_Surface_Conditions  659953 non-null  object 
 10  Road_Type                656159 non-null  object 
 11  Urban_or_Rural_Area      660664 non-null  object 
 12  Weather_Conditions       646551 non-null  object 
 13  Vehicle_Type             660679 non-null  object 
dtypes: f

<h2>DataFrame Basic Descriptive Statistics</h2>

In [4]:
accident.describe()

Latitude      Longitude  Number_of_Casualties  Number_of_Vehicles
count  660654.000000  660653.000000         660679.000000       660679.000000
mean       52.553866      -1.431210              1.357040            1.831255
std         1.406922       1.383330              0.824847            0.715269
min        49.914430      -7.516225              1.000000            1.000000
25%        51.490690      -2.332291              1.000000            1.000000
50%        52.315641      -1.411667              1.000000            2.000000
75%        53.453452      -0.232869              1.000000            2.000000
max        60.757544       1.762010             68.000000           32.000000

<h2>Checking Null Values</h2>

In [5]:
accident.isnull().sum()

Index                          0
Accident_Severity              0
Accident Date                  0
Latitude                      25
Light_Conditions               0
District Area                  0
Longitude                     26
Number_of_Casualties           0
Number_of_Vehicles             0
Road_Surface_Conditions      726
Road_Type                   4520
Urban_or_Rural_Area           15
Weather_Conditions         14128
Vehicle_Type                   0
dtype: int64

<h2>Fill Null Values with Statistical Values</h2>

In [6]:
accident['Latitude'] = accident['Latitude'].fillna(accident['Latitude'].mean())
accident['Longitude'] = accident['Longitude'].fillna(accident['Longitude'].mean())
accident['Road_Surface_Conditions'] = accident['Road_Surface_Conditions'].fillna(accident['Road_Surface_Conditions'].mode()[0])
accident['Road_Type'] = accident['Road_Type'].fillna(accident['Road_Type'].mode()[0])
accident['Urban_or_Rural_Area'] = accident['Urban_or_Rural_Area'].fillna(accident['Urban_or_Rural_Area'].mode()[0])
accident['Weather_Conditions'] = accident['Weather_Conditions'].fillna(accident['Weather_Conditions'].mode()[0])

<h2>Checking Column Data Types</h2>

In [7]:
accident.dtypes

Index                       object
Accident_Severity           object
Accident Date               object
Latitude                   float64
Light_Conditions            object
District Area               object
Longitude                  float64
Number_of_Casualties         int64
Number_of_Vehicles           int64
Road_Surface_Conditions     object
Road_Type                   object
Urban_or_Rural_Area         object
Weather_Conditions          object
Vehicle_Type                object
dtype: object

In [8]:
# Convert to integer
accident['Number_of_Casualties'] = accident['Number_of_Casualties'].astype('int')
accident['Number_of_Vehicles'] = accident['Number_of_Vehicles'].astype('int')

# Convert to float
accident['Latitude'] = accident['Latitude'].astype('float')
accident['Longitude'] = accident['Longitude'].astype('float')

# Convert to string (object)
accident['Index'] = accident['Index'].astype('category')
accident['Accident_Severity'] = accident['Accident_Severity'].astype('category')
accident['Accident Date'] = accident['Accident Date'].astype('category')
accident['Light_Conditions'] = accident['Light_Conditions'].astype('category')
accident['District Area'] = accident['District Area'].astype('category')
accident['Road_Surface_Conditions'] = accident['Road_Surface_Conditions'].astype('category')
accident['Road_Type'] = accident['Road_Type'].astype('category')
accident['Urban_or_Rural_Area'] = accident['Urban_or_Rural_Area'].astype('category')
accident['Weather_Conditions'] = accident['Weather_Conditions'].astype('category')
accident['Vehicle_Type'] = accident['Vehicle_Type'].astype('category')

In [9]:
accident

Index Accident_Severity Accident Date   Latitude  \
0       200701BS64157           Serious      5/6/2019  51.506187   
1       200701BS65737           Serious      2/7/2019  51.495029   
2       200701BS66127           Serious    26-08-2019  51.517715   
3       200701BS66128           Serious    16-08-2019  51.495478   
4       200701BS66837            Slight      3/9/2019  51.488576   
...               ...               ...           ...        ...   
660674  201091NM01760            Slight    18-02-2022  57.374005   
660675  201091NM01881            Slight    21-02-2022  57.232273   
660676  201091NM01935            Slight    23-02-2022  57.585044   
660677  201091NM01964           Serious    23-02-2022  57.214898   
660678  201091NM02142           Serious    28-02-2022  57.575210   

                   Light_Conditions           District Area  Longitude  \
0             Darkness - lights lit  Kensington and Chelsea  -0.209082   
1                          Daylight  Kensington and Chelsea  -0.173647   
2       Darkness - lighting unknown  Kensington and Chelsea  -0.210215   
3                          Daylight  Kensington and Chelsea  -0.202731   
4             Darkness - lights lit  Kensington and Chelsea  -0.192487   
...                             ...                     ...        ...   
660674                     Daylight                Highland  -3.467828   
660675       Darkness - no lighting                Highland  -3.809281   
660676                     Daylight                Highland  -3.862727   
660677       Darkness - no lighting                Highland  -3.823997   
660678                     Daylight                Highland  -3.895673   

        Number_of_Casualties  Number_of_Vehicles Road_Surface_Conditions  \
0                          1                   2                     Dry   
1                          1                   2             Wet or damp   
2                          1                   3                     Dry   
3                          1                   4                     Dry   
4                          1                   2                     Dry   
...                      ...                 ...                     ...   
660674                     2                   1                     Dry   
660675                     1                   1            Frost or ice   
660676                     1                   3            Frost or ice   
660677                     1                   2             Wet or damp   
660678                     1                   1             Wet or damp   

                 Road_Type Urban_or_Rural_Area     Weather_Conditions  \
0       Single carriageway               Urban     Fine no high winds   
1       Single carriageway               Urban  Raining no high winds   
2       Single carriageway               Urban     Fine no high winds   
3       Single carriageway               Urban     Fine no high winds   
4       Single carriageway               Urban     Fine no high winds   
...                    ...                 ...                    ...   
660674  Single carriageway               Rural     Fine no high winds   
660675  Single carriageway               Rural     Fine no high winds   
660676  Single carriageway               Rural     Fine no high winds   
660677  Single carriageway               Rural     Fine no high winds   
660678    Dual carriageway               Rural  Snowing no high winds   

                                Vehicle_Type  
0                                        Car  
1                                        Car  
2                      Taxi/Private hire car  
3       Bus or coach (17 or more pass seats)  
4                              Other vehicle  
...                                      ...  
660674                                   Car  
660675                                   Car  
660676                                   Car  
660677                 Motorcycle over 500cc  
660

<h1>Exploratory Data Analytics</h1>

<h3>Question 1</h3>
<h3>Which road surface condition has the fewest accidents?</h3>

In [10]:
accident['Road_Surface_Conditions'].value_counts()

Road_Surface_Conditions
Dry                     448547
Wet or damp             186708
Frost or ice             18517
Snow                      5890
Flood over 3cm. deep      1017
Name: count, dtype: int64

<h3>Insights No. 1</h3>
<h4>The fewest accidents happened when the road was flooded over 3 cm deep, which makes sense because most drivers avoid driving in deep floodwater, while dry and wet roads have the most accidents since those are the conditions people usually drive in every day.</h4>

<h3>Question 2</h3>
<h3>What is the maximum number of casualties recorded in a single accident?</h3>

In [11]:
accident['Number_of_Casualties'].max()

np.int64(68)

<h3>Insights No. 2</h3>
<h4>The deadliest accident recorded had 68 casualties, showing the severe impact a single incident can cause.</h4>

<h3>Question 3</h3>
<h3>What is the minimum number of vehicles involved in accidents?</h3>

In [12]:
accident['Number_of_Vehicles'].min()

np.int64(1)

<h3>Insights No. 3</h3>
<h4>The smallest number of vehicles in an accident is 1, which represents single-vehicle crashes such as hitting barriers, trees, or overturning without another car involved.</h4>

<h3>Question 4</h3>
<h3>What is the mode (most common value) of accident severity?</h3>

In [13]:
accident['Accident_Severity'].mode()[0]

'Slight'

<h3>Insights No. 4</h3>
<h4>The most common accident severity is 3 (slight accidents), meaning that most reported accidents only result in minor injuries or damage instead of severe or fatal outcomes.</h4>

<h3>Question 5</h3>
<h3>What is the mean accident severity score?</h3>

In [14]:
accident['Accident_Severity_numeric'] = accident['Accident_Severity'].astype('category').cat.codes
accident['Accident_Severity_numeric'].mean()

np.float64(1.8402567661451326)

<h3>Insights No. 5</h3>
<h4>The average accident severity score is 1.84, which means most accidents fall between slight and serious in severity.</h4>

<h3>Question 6</h3>
<h3>What is the average number of casualties per accident?</h3>

In [15]:
accident['Number_of_Casualties'].mean()

np.float64(1.357040257068864)

<h3>Insights No. 6</h3>
<h4>On average, each accident results in about 1 to 2 casualties, showing that most crashes involve only a small number of people.</h4>

<h3>Question 7</h3>
<h3>What is the average number of casualties in accidents on dual carriageways?</h3>

In [50]:
dual = accident[accident['Road_Type'] == 'Dual carriageway']
dual['Number_of_Casualties'].mean()

np.float64(1.4772791277759898)

<h3>Insights No. 7</h3>
<h4>On average, accidents on dual carriageways involve about 1.48 casualties per crash, showing that while these roads are built for smoother traffic, crashes can still cause multiple injuries.</h4>

<h3>Question 8</h3>
<h3>What is the average latitude of all recorded accidents?</h3>

In [21]:
accident['Latitude'].mean()

np.float64(52.553865761110956)

<h3>Insights No. 8</h3>
<h4>The mean latitude is 52.55, which indicates that accidents are most common around the central and northern parts of the UK.</h4>

<h3>Question 9</h3>
<h3>What is the maximum number of vehicles ever involved in an accident?</h3>

In [22]:
accident['Number_of_Vehicles'].max()

np.int64(32)

<h3>Insights No. 9</h3>
<h4>The largest recorded accident involved 32 vehicles, showing that while most crashes involve two cars, pile-ups on busy roads or highways can lead to very large accidents.</h4>

<h3>Question 10</h3>
<h3>What is the most common road type where accidents occur?</h3>

In [24]:
accident['Road_Type'].mode()[0]

'Single carriageway'

<h3>Insights No. 10</h3>
<h4>The most frequent road type for accidents is single carriageway, which makes sense because these roads are the most common in the UK and carry both traffic directions without a divider.</h4>

<h3>Question 11</h3>
<h3>How many accidents happened in each district and road type?</h3>

In [26]:
accident.groupby(['District Area','Road_Type']).size()

District Area  Road_Type         
Aberdeen City  Dual carriageway       246
               One way street          29
               Roundabout              87
               Single carriageway     948
               Slip road               13
                                     ... 
York           Dual carriageway       114
               One way street          61
               Roundabout             176
               Single carriageway    1540
               Slip road                6
Length: 2110, dtype: int64

<h3>Insights No. 11</h3>
<h4>Accidents occur across all road types, but single carriageways consistently record the highest numbers in every district, showing they are the most accident-prone road type.</h4>

<h3>Question 12</h3>
<h3>What is the average number of vehicles in each road surface condition?</h3>

In [27]:
accident.groupby('Road_Surface_Conditions')['Number_of_Vehicles'].mean()

Road_Surface_Conditions
Dry                     1.841631
Flood over 3cm. deep    1.610619
Frost or ice            1.640006
Snow                    1.746859
Wet or damp             1.829161
Name: Number_of_Vehicles, dtype: float64

<h3>Insights No. 12</h3>
<h4>On average, accidents on dry (1.84) and wet/damp (1.83) roads involve slightly more vehicles compared to flooded, icy, or snowy conditions, suggesting that high-traffic normal conditions may lead to multi-vehicle collisions.</h4>

<h3>Question 13</h3>
<h3>What is the average accident severity in each district?</h3>

In [30]:
accident['Accident_Severity_numeric'] = accident['Accident_Severity'].astype('category').cat.codes
accident.groupby('District Area')['Accident_Severity_numeric'].mean()

District Area
Aberdeen City    1.801209
Aberdeenshire    1.691710
Adur             1.810985
Allerdale        1.830674
Alnwick          1.806034
                   ...   
Wychavon         1.814107
Wycombe          1.851938
Wyre             1.825525
Wyre Forest      1.818369
York             1.843437
Name: Accident_Severity_numeric, Length: 422, dtype: float64

<h3>Insights No. 13</h3>
<h4>Some districts have a higher average severity score, meaning accidents there tend to be more serious, while others mostly experience minor crashes.</h4>

<h3>Question 14</h3>
<h3>What is the minimum number of casualties ever recorded in an accident?</h3>

In [51]:
accident['Number_of_Casualties'].min()

np.int64(1)

<h3>Insights No. 14</h3>
<h4>The smallest number of casualties in an accident is 1, which means that every recorded accident in the dataset involved at least one person being injured or affected.</h4>

<h3>Question 15</h3>
<h3>What is the correlation between number of vehicles and casualties?</h3>

In [32]:
accident['Number_of_Vehicles'].corr(accident['Number_of_Casualties'])

np.float64(0.22888886126927557)

<h3>Insights No. 15</h3>
<h4>When more vehicles are involved in an accident, it usually leads to more people getting hurt, but the link isn’t that strong since other things like road conditions, driver actions, and safety measures also make a big difference.</h4>

<h3>Question 16</h3>
<h3>What is the correlation between latitude and casualties?</h3>

In [33]:
accident['Latitude'].corr(accident['Number_of_Casualties'])

np.float64(0.032200686625906395)

<h3>Insights No. 16</h3>
<h4>The north–south location (latitude) of an accident has almost no effect on how many people get hurt, showing that casualties are influenced more by other factors.</h4>

<h3>Question 17</h3>
<h3>What is the correlation between longitude and casualties?</h3>

In [35]:
accident['Longitude'].corr(accident['Number_of_Casualties'])

np.float64(-0.0404056457884545)

<h3>Insights No. 17</h3>
<h4>The east–west location (longitude) of an accident shows almost no link to how many people are hurt, meaning casualties are driven by other factors rather than geography.</h4>

<h3>Question 18</h3>
<h3>Compare casualties in urban and rural areas.</h3>

In [54]:
from scipy.stats import f_oneway
dry = accident[accident['Road_Surface_Conditions'] == 'Dry']['Number_of_Casualties']
wet = accident[accident['Road_Surface_Conditions'] == 'Wet or damp']['Number_of_Casualties']
stats, p_value = f_oneway(dry, wet)
p_value

np.float64(5.097050919646033e-249)

<h3>Insights No. 18</h3>
<h4>Accidents on wet or damp roads lead to significantly different casualty numbers compared to dry roads, showing that slippery conditions greatly affect how severe crashes can be.</h4>

<h3>Question 19</h3>
<h3>Compare casualties between single and dual carriageway roads.</h3>

In [47]:
type1 = accident[accident['Road_Type'] == 'Single carriageway']['Number_of_Casualties']
type2 = accident[accident['Road_Type'] == 'Dual carriageway']['Number_of_Casualties']
stats, p_value = f_oneway(type1, type2)
p_value

np.float64(0.0)

<h3>Insights No. 19</h3>
<h4>Crashes on single carriageway roads usually result in more people getting hurt than those on dual carriageways, highlighting how safer road designs can reduce accident severity.</h4>

<h3>Question 20</h3>
<h3>Create a column for casualties per vehicle.</h3>

In [48]:
accident['Casualties_per_Vehicle'] = accident['Number_of_Casualties'] / accident['Number_of_Vehicles']
accident[['Number_of_Casualties','Number_of_Vehicles','Casualties_per_Vehicle']]

Number_of_Casualties  Number_of_Vehicles  Casualties_per_Vehicle
0                          1                   2                0.500000
1                          1                   2                0.500000
2                          1                   3                0.333333
3                          1                   4                0.250000
4                          1                   2                0.500000
...                      ...                 ...                     ...
660674                     2                   1                2.000000
660675                     1                   1                1.000000
660676                     1                   3                0.333333
660677                     1                   2                0.500000
660678                     1                   1                1.000000

[660679 rows x 3 columns]

<h3>Insights No. 20</h3>
<h4>This new column shows accident intensity by comparing casualties to vehicles involved, which helps us better compare small crashes with large pile-ups.</h4>